memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,757 @@
+presented below:
+
+.. code-block::
+
+   distance_reserved_bits = 3
+   sort(tagged_symbols) # by symbol.address, ascending
+
+   last_symbol_end = 0
+   for symbol in tagged_symbols:
+      distance = ((symbol.address - last_symbol_end) / tag_granule_size)
+                  << distance_reserved_bits
+      size = symbol.size / tag_granule_size
+      last_symbol_end = symbol.address + symbol.size
+
+      if size < (1 << distance_reserved_bits):
+         write_to_section(uleb128(distance | size))
+      else:
+         write_to_section(uleb128(distance))
+         write_to_section(uleb128(size - 1))
+
+.. raw:: pdf
+
+   PageBreak
+
+Relocation Operations
+---------------------
+
+In order to support materialization of the logical tag into the relevant GOT
+entry for a tagged global, some AArch64 relocations have semantics that extend
+the AAELF64_ definition. Because global variable support in the MemtagABI is
+designed to be backwards-compatible, the result of the relocation on a
+non-MemtagABI binary or non-MTE device is the same as the non-extended variant.
+
+The relocations reference the following mnemonics:
+
+- ``LDG(pointer)`` is an instruction for the run-time environment to use the
+  ``ldg`` instruction on ``pointer`` to materialize the correct logical tag for
+  a symbol. This operation should also align the pointer down to the closest tag
+  granule before executing the ``ldg`` instruction.
+
+- ``TAG(pointer, offset)`` is equivalent to ``LDG(pointer + offset) - offset``.
+  This allows for the materialization of the correct logical tag for relocations
+  where the result falls out-of-bounds of the referenced symbol (i.e. the addend
+  for the relocation is negative, or greater than or equal to the size of the
+  symbol). In these instances, the logical tag should come from a a different
+  place than the rest of the resulting pointer.
+
+- For all the relocation types listed below, the loader should use the ``ldg``
+  instruction on the target address before writing into the target field, as the
+  target field may be inside of a tagged region.
+
+.. table:: Relocations with extended semantics
+
+  +------------+--------------------+----------------+-------------------------------------+
+  | ELF64 Code | Name               | Base Operation | MemtagABI Extended Operation        |
+  +============+====================+================+=====================================+
+  | 257	      | R_AARCH64_ABS64    | S + A          | LDG(S) + A                          |
+  +------------+--------------------+----------------+-------------------------------------+
+  | 1025	      | R_AARCH64_GLOB_DAT | S + A          | LDG(S) + A                          |
+  +------------+--------------------+----------------+-------------------------------------+
+  | 1027	      | R_AARCH64_RELATIVE | Delta(S) + A   | TAG(Delta(S) + A, \*(Delta(S) + A)) |
+  +------------+--------------------+----------------+-------------------------------------+
```

Comment:
This strikes me as odd -- can I double-check that my understanding is correct?
If we want a relocation pointing to `foo + 32` where `foo` is only 32 bytes, does this mean that the offset of `-32` must be located at `foo + 32`?
That seems like it would require much more complexity around layout than we have already.

Reply:
Almost, a relocation pointing to `foo + 32` where `foo` is 32 bytes, the `-32` offset is put wherever the relocation result would be written.

More concretely:

```
char a[32]; // &a[0] = 0x1000 (.bss)
char* b = a[32]; // &b = 0x2000 (.data)
```

To relocate `b`, we generate a `R_AARCH64_RELATIVE` relocation for `0x1020 (0x1000 + 32)`. Then, we store the "tag derivation offset" in the place being relocated, i.e. `*0x2000 == -32`. Thus, the tag is loaded from `0x1020 + *0x2000 -> 0x1020 + (-32) -> 0x1000`.

Reply:
That makes a lot more sense, thanks!

IIUC the use of `*(Delta(S) + A)` for the offset argument to `TAG` doesn't describe that.
Since `Delta(S) + A` represents the value that we calculate with the `R_AARCH64_RELATIVE` operation -- which is `foo + 32`.

Would `Delta(0) + <some new marker to represent r_offset>` represent that value better?
(I'm suggesting `Delta(0)` since I believe that makes it explicitly calculate the offset of the place.
Since IIUC even though RELATIVE relocations are usually given the NULL symbol they could be given some other symbol and end up using an offset that may not work for the place we want to store the relocation.)

Reply:
Hmm, seems like `*P` is maybe a better way to describe this (i.e. the offset stored in the place). Given the operation is much more succinct with this, removed the whole `TAG()` operation and inlined it.